ravenframework/Optimizers/GeneticAlgorithm.py

```diff
@@ -320,56 +340,118 @@ def handleInput(self, paramInput):
-
-    # Check if the fitness requested is among the constrained optimization fitnesses
-    # Currently, only InvLin and feasibleFirst Fitnesses deal with constrained optimization
+    
+    ####################################################################################
+    # constraint node                                                                  #
+    ####################################################################################
     # TODO: @mandd, please explore the possibility to convert the logistic fitness into a constrained optimization fitness.
-    if 'Constraint' in self.assemblerObjects and self._fitnessType not in ['invLinear','feasibleFirst','rank_crowding']:
-      self.raiseAnError(IOError, f'Currently constrained Genetic Algorithms only support invLinear and feasibleFirst fitnesses, whereas provided fitness is {self._fitnessType}')
+    if 'Constraint' in self.assemblerObjects and self._fitnessType not in ['invLinear','logistic', 'feasibleFirst','hardConstraint']:
+      self.raiseAnError(IOError, f'Currently constrained Genetic Algorithms only support invLinear, logistic, feasibleFirst and hardConstraint as a fitness, whereas provided fitness is {self._fitnessType}')
+    self._expConstr = self.assemblerObjects['Constraint'][0] if 'Constraint' in self.assemblerObjects else None
+    self._impConstr = self.assemblerObjects['ImplicitConstraint'][0] if 'ImplicitConstraint' in self.assemblerObjects else None
+    if self._expConstr != None and self._impConstr != None:
+      self._numOfConst = len([ele for ele in self._expConstr if ele != 'Functions' if ele !='External']) + len([ele for ele in self._impConstr if ele != 'Functions' if ele !='External']) 
+    elif self._expConstr == None and self._impConstr != None:
+      self._numOfConst = len([ele for ele in self._impConstr if ele != 'Functions' if ele !='External']) 
+    elif self._expConstr != None and self._impConstr == None:
+      self._numOfConst = len([ele for ele in self._expConstr if ele != 'Functions' if ele !='External']) 
+    else:
+      self._numOfConst = 0
+    if (self._expConstr != None) and (self._impConstr != None) and (self._penaltyCoeff != None):
+      if len(self._penaltyCoeff) != len(self._objectiveVar) * self._numOfConst:
+        self.raiseAnError(IOError, f'The number of penaltyCoeff. in <b> should be identical with the number of objective in <objective> and the number of constraints (i.e., <Constraint> and <ImplicitConstraint>)')
+    else:
+      pass
     self._objCoeff = fitnessNode.findFirst('a').value if fitnessNode.findFirst('a') is not None else None
-    self._penaltyCoeff = fitnessNode.findFirst('b').value if fitnessNode.findFirst('b') is not None else None
+    #NOTE the code lines below are for 'feasibleFirst' temperarily. It will be generalized for invLinear as well.
+    if self._fitnessType == 'feasibleFirst':  
+      if self._numOfConst != 0 and fitnessNode.findFirst('b') is not None:
```

Comment:
Inside GeneticAlgorithms.py, we stripped out fitness, so we do not want to have several if statements for each fitness type. Can't we move this to the fitness.py under feasibleFirst?

Reply:
Unless it is required, I would like to suggest letting these lines under "handleInput" method.